keps/sig-network/3458-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,787 @@
+B) as well. The health check for class B) services should hence answer if
+endpoints are running on the node AND if Kube-proxy is healthy.
+
+[1]: https://github.com/kubernetes/kubernetes/blob/f9afd68e3b3fc05f3607ea7cb90808c3e931ba9c/pkg/proxy/apis/config/validation/validation.go#L38
+
+#### Defining what "healthy" should be for Kube-proxy
+
+This section describes how Kube-proxy might handle covering all above cases.
+
+As mentioned above, A) is not handled by Kube-proxy today.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+I don't have any good ideas on how to solve it either...running the HC server in the Kubelet instead?
+<<[/UNRESOLVED]>>
+```
+
+B) shows how Kube-proxy today plays in a good way and correctly handles the
+disaster case of: "all Kube-proxies cannot read from the API server". It also
+goes to show that changes need to be made as to handle: "some Kube-proxies
+cannot read from the API server". api-machinery currently provides informers a
+mechanism for intercepting watch errors and reacting to them, which Kube-proxy
+currently does not use. The function `SetWatchErrorHandler` allows the caller to
+define a callback which is invoked when errors are intercepted, below is an
+example of how it can be used with the `DefaultWatchErrorHandler`:
+
+```go
+
+func DefaultWatchErrorHandler(r *Reflector, err error) {
+	switch {
+	case isExpiredError(err):
+		// Don't set LastSyncResourceVersionUnavailable - LIST call with ResourceVersion=RV already
+		// has a semantic that it returns data at least as fresh as provided RV.
+		// So first try to LIST with setting RV to resource version of last observed object.
+		klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
+	case err == io.EOF:
+		// watch closed normally
+	case err == io.ErrUnexpectedEOF:
+		klog.V(1).Infof("%s: Watch for %v closed with unexpected EOF: %v", r.name, r.expectedTypeName, err)
+	default:
+		utilruntime.HandleError(fmt.Errorf("%s: Failed to watch %v: %v", r.name, r.expectedTypeName, err))
+	}
+}
+
+serviceInformer.Informer().SetWatchErrorHandler(DefaultWatchErrorHandler)
+```
+
+The main complexity arising from trying to handle this case comes from
+differentiating correctly between "master has a problem" and "I have a problem
+reading from master". Unless a combination of [the defined status messages](2) can be
+crafted to correctly differentiate between those two modes, the only mechanism
+would be to run a consensus loop between all Kube-proxy members to establish
+this. Doing that might however introduce more complexity than it tries to solve.
+If a single Kube-proxy replica encounters issues reading from the API server
+because of incorrect credentials, then there might be status codes which any
+Kube-proxy instance could use to correctly indicate that "only it has a
+problem", ex: `StatusReasonUnauthorized` / `StatusReasonForbidden`.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
```

Comment:
I think this wants to be  KEP of it's own, too.

We could do something like a per-node lease for kube-proxy.  Every k-p needs to watch every such lease, but then we can make local decisions about scope of a failure.

Reply:
> I think this wants to be KEP of it's own

Yes, and maybe some API server folks might have some good pointers for dealing with consensus related failure modes like this?

Reply:
or we end up inventing a solution that other node-agents would copy.

Reply:
punted to a later KEP